packages/data-point/src/ReducerList.js

```diff
@@ -1,42 +1,46 @@
-const { ReducerNative } = require("./ReducerNative");
+const { Reducer } = require("./Reducer");
 
-class ReducerList extends ReducerNative {
+class ReducerList extends Reducer {
   constructor(spec, createReducer) {
     super("list", undefined, spec);
 
-    this.reducerList = spec.map(token => createReducer(token));
+    this.reducerList = spec.map(reducerSource => createReducer(reducerSource));
   }
 
   static isType(spec) {
     return Array.isArray(spec);
   }
 
   async resolve(accumulator, resolveReducer) {
-    const reducers = this.reducerList;
+    const reducerList = this.reducerList;
 
-    if (reducers.length === 0) {
+    if (reducerList.length === 0) {
       return undefined;
     }
 
-    const initialValue =
-      accumulator.value === undefined ? null : accumulator.value;
+    let value = accumulator.value;
+    let acc = accumulator;
 
     let index = 0;
+    let hasReachedEnd;
+    const reducerListLength = reducerList.length;
 
-    let value = initialValue;
-
-    while (index < reducers.length) {
-      const reducer = reducers[index];
-
-      const acc = Object.create(accumulator);
-      acc.value = value;
+    do {
+      const reducer = reducerList[index];
 
       // we do purposely want to wait for each reducer to execute
       // eslint-disable-next-line no-await-in-loop
       value = await resolveReducer(acc, reducer);
 
       index += 1;
-    }
+
+      hasReachedEnd = index === reducerListLength;
+
+      if (!hasReachedEnd) {
+        // only create new accumulator if there are more reducers to process
+        acc = acc.set("value", value);
+      }
+    } while (!hasReachedEnd);
```

Comment:
this can be `while (!hasReachedEnd) {}` since `hasReachedEnd` always starts as undefined

Reply:
actually, how about a for loop?
- it communicates that we're visiting every array element (in order)
- it's a little more self-contained because variables like `hasReachedEnd` are not defined outside of the loop block

```js
const maxIndex = reducerList.length - 1
for (let i = 0; i <= maxIndex; i++) {
  const reducer = reducerList[i];
  // we do purposely want to wait for each reducer to execute
  // eslint-disable-next-line no-await-in-loop
  value = await resolveReducer(acc, reducer);
  if (i < maxIndex) {
    // only create new accumulator if there are more reducers to process
    acc = acc.set("value", value);
  }
}
```

Reply:
thanks! had my head looped in here, don't know how I ended up on a do{}while :)